### PR TITLE
Change slider value on mouse up on bar

### LINKED ISSUE #4233: Clicking on Slider (but not on handle) should change the Slider's value
Now, I click slider, the value won't change, and I want to know the event like `onRelease` when mouse click released, How can I solve this.

Edit: Acceptance criteria is that when clicking within a slider's geometry but _not_ on the handle, the value should change according to platform defaults. Typically a click means a moving the knob to the location of the mouse (as if it was dragged all the way), and adjusting the value accordingly.

### PR DESCRIPTION
**Pull request by FloVanGH**
Fixes #4233 

This also changes the behavior of `Slider`. Mouse-Left/Touch down on the bar outside the thumb adjust the value on the mouse position. Drag move only starts if the pointer is inside the thumb.

### CODE REVIEW

File: internal/compiler/widgets/common/slider-base.slint [hunk 1]
```diff
@@ -3,43 +3,51 @@
...
             if (event.button == PointerEventButton.left && event.kind == PointerEventKind.down) {
                 self.pressed-value = root.value;
+                self.pointer-moved = false;
+                return;
+            }
+
+            if (!self.pointer-moved && event.button == PointerEventButton.left && event.kind == PointerEventKind.up) {
+                 root.set-value((!root.vertical ? root.size-to-value(touch-area.mouse-x, root.width) :
+                    root.size-to-value(touch-area.mouse-y, root.height)) + root.minimum);
+                self.pointer-moved = false;
```
Comment (tronical) [inline internal/compiler/widgets/common/slider-base.slint:37]: I'm curious, why is `pointer-moved` set to false here?

**Review by tronical — COMMENTED** [review]
Oddly, it doesn't work for me on macOS. I recorded a short clip that also visualises the mouse clicks:

https://github.com/slint-ui/slint/assets/1486/930ec0a5-6df6-4e71-942d-a591c42c6156

What am I missing?

File: internal/compiler/widgets/common/slider-base.slint (hunk 1 above)
Comment (FloVanGH) [reply internal/compiler/widgets/common/slider-base.slint:37]: To cleanup the state. But now I am thinking about it, I think to set it by `down` should be enough.

**Comment by FloVanGH** [comment]
> Oddly, it doesn't work for me on macOS. I recorded a short clip that also visualises the mouse clicks:
> Screen.Recording.2024-01-23.at.17.16.05.mov
> 
> What am I missing?

Strange. It works for me on Wayland (Gnome). I though I tested it also on macOS. I will check it again.

**Comment by FloVanGH** [comment]
> > Oddly, it doesn't work for me on macOS. I recorded a short clip that also visualises the mouse clicks:
> > Screen.Recording.2024-01-23.at.17.16.05.mov
> > What am I missing?
> 
> Strange. It works for me on Wayland (Gnome). I though I tested it also on macOS. I will check it again.

@tronical it is really strange, it works perfectly fine for me on macOS. It works on each mouse left button up.

**Comment by tronical** [comment]
I debugged this a little bit and for me the mouse pointer is moved (by very few pixels) if I click.

Ideally, a widget should not have to do special tracking like that. Sadly, `clicked => {}` can't be used because that will be invoked regardless of how many pixels the pointer is moved, and we probably can't change that behaviour (and shouldn't?).
But perhaps this is a good reason to expose the click count, because that accounts for minimal movement. @ogoffart what do you think?

**Comment by ogoffart** [comment]
> But perhaps this is a good reason to expose the click count, because that accounts for minimal movement. @ogoffart what do you think?

I don't think that helps in that case because the click count is not reset on release (maybe it should?  And it also starts at 0

**Comment by FloVanGH** [comment]
@tronical @ogoffart I think there is an alternative to fix this situation. Common behavior for Slider is that it can only drag moved on the slider thumb and not on by press move on the whole bar. This could help. What do you think? But this would be a change of the Sliders behavior

**Comment by ogoffart** [comment]
But what does the native slider does?
The Qt slider on KDE seems like it doesn't care if you move the mouse outside of the handle.

**Comment by tronical** [comment]
> > But perhaps this is a good reason to expose the click count, because that accounts for minimal movement. @ogoffart what do you think?
> 
> I don't think that helps in that case because the click count is not reset on release (maybe it should? And it also starts at 0

If it wouldn't start at 0 (and we can change that since we haven't exposed it yet), wouldn't it suffice to write `if (event.click-count > 0 && mod(event.click-count, 2) == 0)` ?

**Comment by FloVanGH** [comment]
> But what does the native slider does? The Qt slider on KDE seems like it doesn't care if you move the mouse outside of the handle.

On long run I would prefer to change the behavior and to do drag move only on the thumb. But I think it is ok now to leave this behavior now as it is, until we have a working solution for the click e.g. with the count.

**Comment by tronical** [comment]
> But what does the native slider does? The Qt slider on KDE seems like it doesn't care if you move the mouse outside of the handle.

I recorded three clips of native sliders:

Windows, in the Windows System Settings app:


https://github.com/slint-ui/slint/assets/1486/8d68c66c-3585-479f-a7d4-123fd8661427

macOS, in the macOS control panel:


https://github.com/slint-ui/slint/assets/1486/372c741b-b3ab-4eb2-85f3-2cd99ca88440

In Gnome, on the Gnome system settings app:

[gnome.webm](https://github.com/slint-ui/slint/assets/1486/4326b0f7-80fd-416c-80cf-c113fd7df67e)


In these three instances, clicking on the area that is _not_ the handle moves it to that location (with and without animation).

**Comment by FloVanGH** [comment]
> > But what does the native slider does? The Qt slider on KDE seems like it doesn't care if you move the mouse outside of the handle.
> 
> I recorded three clips of native sliders:
> 
> Windows, in the Windows System Settings app:
> windows.mov
> 
> macOS, in the macOS control panel:
> mac.mov
> 
> In Gnome, on the Gnome system settings app:
> gnome.webm
> 
> In these three instances, clicking on the area that is _not_ the handle moves it to that location (with and without animation).

And drag move works only on the handle, right?

**Comment by tronical** [comment]
And one more from KDE:

https://github.com/slint-ui/slint/assets/1486/7b974ae3-589d-4549-a9d0-64d960a32103

**Comment by tronical** [comment]
Sounds good to me. I agree, we're approximating here. Thanks for improving the behaviour :)

**Comment by tronical** [comment]
For the record, even if this doesn't work on my Mac yet because on my high-res screen the mouse still "moves", I think doing this approach is an improvement and worth doing. FWIW, it works on my dell laptop with linux - a clear improvement there :)

**Review by tronical — APPROVED** [review]
Works great for me (now also on macOS, as not sensitive to the move :)

**Pull request merged**